pkg/awsclient/ec2_helper.go

```diff
@@ -0,0 +1,188 @@
+					{
+						CidrIp:      aws.String(cidrBlock),
+						Description: aws.String("Approved CIDR Block from cloud-ingress-operator configuration"),
+					},
+				},
+			})
+		}
+	}
+	if err := c.addIngressRulesToSecurityGroup(securityGroup, rulesToAdd); err != nil {
+		return err
+	}
+	if err := c.removeIngressRulesFromSecurityGroup(securityGroup, rulesToRemove); err != nil {
+		return err
+	}
+	// Once the ingress rules are updated, attach the SecurityGroup to the load balancer
+	return c.setLoadBalancerSecurityGroup(loadBalancerName, securityGroup)
+}
+
+// Add rules to the security group
+func (c *awsClient) addIngressRulesToSecurityGroup(securityGroup *ec2.SecurityGroup, ipPermissions []*ec2.IpPermission) error {
+	if len(ipPermissions) == 0 {
+		// nothing to do
+		return nil
+	}
+	i := &ec2.AuthorizeSecurityGroupIngressInput{
+		IpPermissions: ipPermissions,
+		GroupId:       securityGroup.GroupId,
+	}
+	_, err := c.AuthorizeSecurityGroupIngress(i)
+	return err
+}
+
+// Remove rules from the security group
+func (c *awsClient) removeIngressRulesFromSecurityGroup(securityGroup *ec2.SecurityGroup, ipPermissions []*ec2.IpPermission) error {
+	if len(ipPermissions) == 0 {
+		// nothing   to do
+		return nil
+	}
+	i := &ec2.RevokeSecurityGroupIngressInput{
+		FromPort:      aws.Int64(6443),
+		ToPort:        aws.Int64(6443),
+		IpProtocol:    aws.String("tcp"),
+		IpPermissions: ipPermissions,
+		GroupId:       securityGroup.GroupId,
+	}
+	_, err := c.RevokeSecurityGroupIngress(i)
+	return err
+}
+
+// createSecurityGroup creates a SecurityGroup with the given name, and returns the EC2 object and/or any error
+func (c *awsClient) createSecurityGroup(securityGroupName, vpcId string) (*ec2.SecurityGroup, error) {
+	createInput := &ec2.CreateSecurityGroupInput{
+		Description: aws.String("Admin API Security group"),
+		GroupName:   aws.String(securityGroupName),
+		VpcId:       aws.String(vpcId),
+	}
+	createResult, err := c.CreateSecurityGroup(createInput)
+	if err != nil {
+		return nil, err
+	}
+
+	// Caller of this method wants a *ec2.SecurityGroup, and since the create
+	// method doesn't give us nought but the group-id, we have to do a search
+	// to find it.
+	return c.findSecurityGroupByID(*createResult.GroupId)
+}
+
+func (c *awsClient) findSecurityGroupByID(id string) (*ec2.SecurityGroup, error) {
+	i := &ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("group-id"),
+				Values: aws.StringSlice([]string{id}),
+			},
+		},
+	}
+	o, err := c.DescribeSecurityGroups(i)
+	if err != nil {
+		return nil, err
+	}
+	if len(o.SecurityGroups) == 0 {
+		return nil, nil
+	}
+	return o.SecurityGroups[0], nil
+}
+
+func (c *awsClient) findSecurityGroupByName(name string) (*ec2.SecurityGroup, error) {
+	i := &ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("group-name"),
+				Values: aws.StringSlice([]string{name}),
+			},
+		},
+	}
+	o, err := c.DescribeSecurityGroups(i)
```

Comment:
Does this return an err if the group doesn't exist? If so, then line 174 would never be called.

Reply:
It does not return an error.

Reply:
Documentation on that?

Reply:
The behaviour in [the documentation for this method](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeSecurityGroups) does not indicate that any error is returned when there are no matches in the specified filters. Additionally, the [common errors](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html#CommonErrors) for the library fail to indicate this occurrence. Granted, the absence of this documentation doesn't mean that it doesn't happen, but it is hard to prove a negative. ;)

As for anecdotal evidence, I have a small test program I used to back up my statement.